rpcs3/Emu/RSX/Program/GLSLSnippets/RSXProg/RSXVertexFetch.glsl

```diff
@@ -0,0 +1,209 @@
+R"(
+#ifdef _INTEL_GLSL
+// For intel GPUs which cannot access vectors in indexed mode (driver bug? or glsl version too low?)
+// Note: Tested on Mesa iris with HD 530 and compilant path works fine, may be a bug on Windows proprietary drivers
+void mov(inout uvec4 vector, const in int index, const in uint scalar)
+{
+	switch(index)
+	{
+		case 0: vector.x = scalar; return;
+		case 1: vector.y = scalar; return;
+		case 2: vector.z = scalar; return;
+		case 3: vector.w = scalar; return;
+	}
+}
+
+uint ref(const in uvec4 vector, const in int index)
+{
+	switch(index)
+	{
+		case 0: return vector.x;
+		case 1: return vector.y;
+		case 2: return vector.z;
+		case 3: return vector.w;
+	}
+}
+#else
+#define mov(v, i, s) v[i] = s
+#define ref(v, i) v[i]
+#endif
+
+#ifdef VULKAN
+#define _gl_VertexID gl_VertexIndex
+#else
+#define _gl_VertexID gl_VertexID
+#endif
+
+struct attribute_desc
+{
+	uint type;
+	uint attribute_size;
+	uint starting_offset;
+	uint stride;
+	uint frequency;
+	bool swap_bytes;
+	bool is_volatile;
+	bool modulo;
+};
+
+uint gen_bits(const in uint x, const in uint y, const in uint z, const in uint w, const in bool swap)
+{
+	return (swap) ?
+		_set_bits(_set_bits(_set_bits(w, z, 8, 8), y, 16, 8), x, 24, 8) :
+		_set_bits(_set_bits(_set_bits(x, y, 8, 8), z, 16, 8), w, 24, 8);
+}
+
+uint gen_bits(const in uint x, const in uint y, const in bool swap)
+{
+	return (swap)? _set_bits(y, x, 8, 8) : _set_bits(x, y, 8, 8);
+}
+
+// NOTE: (int(n) or int(n)) is broken on some NVIDIA and INTEL hardware when the sign bit is involved.
+// See https://github.com/RPCS3/rpcs3/issues/8990
+vec4 sext(const in ivec4 bits)
+{
+	// convert raw 16 bit values into signed 32-bit float4 counterpart
+	bvec4 sign_check = lessThan(bits, ivec4(0x8000));
+	return _select(bits - 65536, bits, sign_check);
+}
+
+float sext(const in int bits)
+{
+	return (bits < 0x8000) ? float(bits) : float(bits - 65536); 
+}
+
+vec4 fetch_attribute(const in attribute_desc desc, const in int vertex_id, usamplerBuffer input_stream)
+{
+	const int elem_size_table[] = { 0, 2, 4, 2, 1, 2, 4, 1 };
+	const float scaling_table[] = { 1., 32767.5, 1., 1., 255., 1., 32767., 1. };
+	const int elem_size = elem_size_table[desc.type];
+	const vec4 scale = scaling_table[desc.type].xxxx;
+
+	uvec4 tmp, result = uvec4(0u);
+	vec4 ret;
+	int n, i = int((vertex_id * desc.stride) + desc.starting_offset);
+
+	for (n = 0; n < desc.attribute_size; n++)

```

Comment:
n seems unused past the loop, so why not put it in the for declaration ?

Reply:
Because intel's awful compiler used to crash when doing that (it's also why the awful INTEL_GLSL section exists). Probably fixed now, but we gain nothing by moving it. There is work slated to deprecate intel opengl compiler entirely and only support SPIRV shaders on that platform, but for now we just have to keep it.